integration_tests/src/main/python/window_function_test.py

```diff
@@ -1066,3 +1066,24 @@ def test_unbounded_to_unbounded_window():
     assert_gpu_and_cpu_are_equal_collect(lambda spark : spark.range(1024).selectExpr(
         'SUM(id) OVER ()',
         'COUNT(1) OVER ()'))
+
+_nested_gens = array_gens_sample + struct_gens_sample + map_gens_sample
+@pytest.mark.parametrize('data_gen', all_basic_gens_no_null + decimal_gens + _nested_gens, ids=idfn)
+def test_window_first_last(data_gen):
+    assert_gpu_and_cpu_are_equal_sql(
+        # Coalesce is to make sure that first and last, which are non-deterministic become deterministic
+        lambda spark: three_col_df(spark, data_gen, string_gen, int_gen, num_slices=1).coalesce(1),
+        "window_agg_table",
+        'SELECT '
+            'a, b, c, '
+            'first(a) OVER (PARTITION BY b ORDER BY c RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW), '
+            'first(a) OVER (PARTITION BY b ORDER BY c RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING), '
+            'first(a) OVER (PARTITION BY b ORDER BY c ROWS  BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW), '
+            'first(a) OVER (PARTITION BY b ORDER BY c ROWS  BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING), '
+            'last (a) OVER (PARTITION BY b ORDER BY c RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW), '
+            'last (a) OVER (PARTITION BY b ORDER BY c RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING), '
+            'last (a) OVER (PARTITION BY b ORDER BY c ROWS  BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW), '
+            'last (a) OVER (PARTITION BY b ORDER BY c ROWS  BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING), '
+            'first(a, true) OVER (PARTITION BY b ORDER BY c), '
+            'last (a, true) OVER (PARTITION BY b ORDER BY c) '
```

Comment:
Agreeing with @revans2: Let us also add `NTH_VALUE`. E.g. `NTH_VALUE(3, a)`.

Reply:
Ah, this might need a little modification: It looks like SparkSQL changed the supported grammar for this. The `IGNORE NULLS` clause is not supported for `FIRST()` and `LAST()` in Spark 3.1.x, although it is available for `NTH_VALUE()`. :angry: 
It is available for all three functions in 3.2.x. 

A more portable phrasing would be:
```sql
  FIRST(a, true) OVER (...),
  LAST(a, true)  IGNORE NULLS OVER (...),
  NTH_VALUE(a, 3) IGNORE NULLS OVER (...)
```

Reply:
Tested, Spark 3.1.x does not support:

```
select first(a) ignore nulls from values (1,1), (2,2) tab(a,b)
select first(a, true) ignore nulls from values (1,1), (2,2) tab(a,b)

``````

Added cases for `first(a) ignore nulls`.

But for `first(a, true) ignore nulls`, the first `true` already means `ignore null', it's duplicated, so did not add this as a case.
Is it necessary to add this case in?

Reply:
No, sir. You're right. `FIRST(a, true)` is the same as `IGNORE NULLS`. Sorry for the confusion.

I tried to clarify in [my previous comment](https://github.com/NVIDIA/spark-rapids/pull/6000#discussion_r921496669): Let us keep `FIRST(a, true)` and skip `IGNORE NULLS` for the current release. We can switch to to use `IGNORE NULLS` uniformly when we drop support for Spark 3.1.x.

Reply:
Oh, excellent! I see what you've done: You're testing all forms of the query, depending on the release version.